include/crab/domains/symbolic_variable_eq_domain.hpp

```diff
@@ -0,0 +1,1031 @@
+    for (auto &kv : m_classes) {
+      const element_t &key = kv.first;
+      const equivalence_class_t &ec = kv.second;
+      auto dom_ptr = ec.get_absval();
+      if (dom_ptr && dom == *dom_ptr) {
+        return get_all_members_from_an_equiv_class(key);
+      }
+    }
+    return boost::none;
+  }
+
+  /// @brief get the domain value stored in current equivalent class
+  /// @param v an element in some set
+  /// @return null if !contains(x), otherwise, returns a shared pointer to value
+  std::shared_ptr<const domain_t> get(const element_t &x) const {
+    if (is_bottom()) {
+      CRAB_ERROR("called ", domain_name(), "::", __func__, " on bottom");
+    }
+    if (is_top() || !contains(x)) {
+      return nullptr;
+    }
+
+    element_t rep_x = find(x);
+    const equivalence_class_t &ec_x = m_classes.at(rep_x);
+    return ec_x.get_absval();
+  }
+
+  /// @brief Add y into the equivalence class of x
+  /// @param x an element may exists
+  /// @param y an element may exists
+  /// @note  if x does not exist, nothing changed.
+  ///        if y \in some cls, forget it from cls, add y into x's class
+  void add(const element_t &x, const element_t &y) {
+    if (is_bottom() || is_top() || x == y) {
+      return;
+    }
+    if (!contains(x)) {
+      set(x, symbolic_variable_equiality_domain_impl::make_fresh_var_symbol<
+                 symbolic_variable_equiality_domain_impl::symbolic_var>());
+    }
+    if (contains(y)) {
+      *this -= y;
+    }
+    element_t rep_x = find(x);
+    m_parents.insert({y, rep_x});
+  }
+  /**------------------ End union find APIs ------------------**/
+
+  /**------------------ Begin domain APIs ------------------**/
+  // empty union-find
+  symbolic_variable_equiality_domain(
+      lattice_val val = lattice_val::neither_top_nor_bot)
+      : m_val(val) {}
+  symbolic_variable_equiality_domain(const this_domain_t &o) = default;
+
+  symbolic_variable_equiality_domain(this_domain_t &&o) = default;
+  this_domain_t &operator=(const this_domain_t &o) = default;
+  this_domain_t &operator=(this_domain_t &&o) = default;
+
+  bool is_bottom() const { return m_val == lattice_val::bottom; }
+
+  bool is_top() const { return m_val == lattice_val::top; }
+
+  void set_to_top() {
+    clear();
+    m_val = lattice_val::top;
+  }
+
+  void set_to_bottom() {
+    clear();
+    m_val = lattice_val::bottom;
+  }
+
+  void set_neither_top_or_bottom() { m_val = lattice_val::neither_top_nor_bot; }
+
+  bool operator<=(const this_domain_t &e) const {
+    SVEQ_DOMAIN_SCOPED_STATS(".leq");
+    if (is_bottom() || e.is_top()) {
+      // _|_ <= e || this <= top
+      return true;
+    } else if (is_top() || e.is_bottom()) {
+      // top >= e || this >= _|_
+      return false;
+    }
+    // Return true if *this is a refined partitioning of o
+    //    forall cls_x \in this. exists cls_y \in e ::
+    //      cls_y <= cls_x
+    equivalence_class_elems_t left_equiv_classes =
```

Comment:
This is not efficient.

The L <= R operation should check that all equalities that exist in R, exist in L as well (but L may have additional equalities). So you do not need to build equivalence classes of L, once you have an equivalence class of R, simply check that these equalities exists in L.

As soon as you find the first equality from R that is not in L, the answer to <= is false, you can return false and do not compute for the rest of the classes.

With this, you can avoid building all equivalence classes for left and right.